### DISCUSSION TITLE
Toolset 'clang' does not appear to support C++11

### DISCUSSION THREAD

**Discussion by yangjinqi123321** [carla-simulator/carla#7472] [Q&A]
ubuntu 22 run  make launch error

**Comment by xavisolesoft** [comment]
Could you share the log of the error you get running `make launch`, please?

**Reply by yangjinqi123321** [reply]
2024-04-19 19:22:04 (1.17 MB/s) - ‘boost_1_80_0.tar.gz.19’ saved [136670223/136670223]

Elapsed Time downloading from boost carla backup in backblaze: 123 seconds
Setup.sh: Extracting boost for Python 3.
Elapsed Time Extracting boost for Python: 3 seconds
Building B2 engine..

A C++11 capable compiler is required for building the B2 engine.
Toolset 'clang' does not appear to support C++11.

> clang++ -x c++ -std=c++11  check_cxx11.cpp
check_cxx11.cpp:14:10: fatal error: 'thread' file not found
#include <thread>
         ^~~~~~~~
1 error generated.


** Note, the C++11 capable compiler is _only_ required for building the B2
** engine. The B2 build system allows for using any C++ level and any other
** supported language and resource in your projects.


You can specify the toolset as the argument, i.e.:
    ./build.sh [options] gcc

Toolsets supported by this script are:
    acc, clang, como, gcc, intel-darwin, intel-linux, kcc, kylix, mipspro,
    pathscale, pgi, qcc, sun, sunpro, tru64cxx, vacpp

For any toolset you can override the path to the compiler with the '--cxx'
option. You can also use additional flags for the compiler with the
'--cxxflags' option.

A special toolset; cxx, is available which is used as a fallback when a more
specific toolset is not found and the cxx command is detected. The 'cxx'
toolset will use the '--cxx' and '--cxxflags' options, if present.

Options:
    --help                  Show this help message.
    --verbose               Show messages about what this script is doing.
    --debug                 Build b2 with debug information, and no
                            optimizations.
    --guess-toolset         Print the toolset we can detect for building.
    --cxx=CXX               The compiler exec to use instead of the detected
                            compiler exec.
    --cxxflags=CXXFLAGS     The compiler flags to use in addition to the
                            flags for the detected compiler.



Failed to build B2 build engine
make: *** [Util/BuildTools/Linux.mk:142: setup] Error 1

**Reply by xavisolesoft** [reply]
Do you have g++ installed?

g++ is requiered for boost b2 build, probably this is why your build is failing.

**Reply by xieran1994** [reply]
Hi I got the same error with g++ installed, have you solved this issue?

**Reply by smp4966** [reply]
did you find a solution to that . if yes your help would be appreciated

**Reply by jeromepatel** [reply]
I am having the same issue, even with g++ installed, is there any solution to that ?

**Reply by donghaiwang** [reply]
same issue, `g++ -v`  gcc version 10.5.0

**Reply by Mr0Ctrl** [reply]
> 2024-04-19 19:22:04 (1.17 MB/s) - ‘boost_1_80_0.tar.gz.19’ saved [136670223/136670223]
> 
> Elapsed Time downloading from boost carla backup in backblaze: 123 seconds Setup.sh: Extracting boost for Python 3. Elapsed Time Extracting boost for Python: 3 seconds Building B2 engine..
> 
> A C++11 capable compiler is required for building the B2 engine. Toolset 'clang' does not appear to support C++11.
> 
> > clang++ -x c++ -std=c++11  check_cxx11.cpp
> > check_cxx11.cpp:14:10: fatal error: 'thread' file not found
> > #include 
> > ^~~~~~~~
> > 1 error generated.
> 
> ** Note, the C++11 capable compiler is _only_ required for building the B2 ** engine. The B2 build system allows for using any C++ level and any other ** supported language and resource in your projects.
> 
> You can specify the toolset as the argument, i.e.: ./build.sh [options] gcc
> 
> Toolsets supported by this script are: acc, clang, como, gcc, intel-darwin, intel-linux, kcc, kylix, mipspro, pathscale, pgi, qcc, sun, sunpro, tru64cxx, vacpp
> 
> For any toolset you can override the path to the compiler with the '--cxx' option. You can also use additional flags for the compiler with the '--cxxflags' option.
> 
> A special toolset; cxx, is available which is used as a fallback when a more specific toolset is not found and the cxx command is detected. The 'cxx' toolset will use the '--cxx' and '--cxxflags' options, if present.
> 
> Options: --help Show this help message. --verbose Show messages about what this script is doing. --debug Build b2 with debug information, and no optimizations. --guess-toolset Print the toolset we can detect for building. --cxx=CXX The compiler exec to use instead of the detected compiler exec. --cxxflags=CXXFLAGS The compiler flags to use in addition to the flags for the detected compiler.
> 
> Failed to build B2 build engine make: *** [Util/BuildTools/Linux.mk:142: setup] Error 1

I am getting the same error.

I can bypass this error by changing the file _carla/Util/BuildTools/Setup.sh_ and forcing gcc to use "boost" during compilation. However, I thought it was a wrong choice because I was getting more errors later on.

```
    BOOST_TOOLSET="clang-10.0" # here
    BOOST_CFLAGS="-fPIC -std=c++14 -DBOOST_ERROR_CODE_HEADER_ONLY"

    py3="/usr/bin/env python${PY_VERSION}"
    py3_root=`${py3} -c "import sys; print(sys.prefix)"`
    pyv=`$py3 -c "import sys;x='{v[0]}.{v[1]}'.format(v=list(sys.version_info[:2]));sys.stdout.write(x)";`
    ./bootstrap.sh \
        --with-toolset=clang \ # and here
        --prefix=../boost-install \
        --with-libraries=python,filesystem,system,program_options \
        --with-python=${py3} --with-python-root=${py3_root}
```

**Comment by PabloVD** [comment]
Possible workarounds for this problem are discussed on [this issue](https://github.com/carla-simulator/carla/issues/6901).
